// component-playground/lib/components/preview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require("react-dom");

var _server = require("react-dom/server");

var _server2 = _interopRequireDefault(_server);

var _babelStandalone = require("babel-standalone");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* eslint-disable max-statements */


var Preview = function (_Component) {
  _inherits(Preview, _Component);

  function Preview() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Preview);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Preview.__proto__ || Object.getPrototypeOf(Preview)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      error: null
    }, _this._compileCode = function () {
      var _this$props = _this.props,
          code = _this$props.code,
          context = _this$props.context,
          noRender = _this$props.noRender,
          scope = _this$props.scope;

      var generateContextTypes = function generateContextTypes(c) {
        return "{ " + Object.keys(c).map(function (val) {
          return val + ": PropTypes.any.isRequired";
        }).join(", ") + " }";
      };

      var scopeWithProps = _extends({}, scope, { PropTypes: _propTypes2.default });

      if (noRender) {
        return (0, _babelStandalone.transform)("\n        ((" + Object.keys(scopeWithProps).join(", ") + ", mountNode) => {\n          class Comp extends React.Component {\n\n            getChildContext() {\n              return " + JSON.stringify(context) + ";\n            }\n\n            render() {\n              return (\n                " + code + "\n              );\n            }\n          }\n\n          Comp.childContextTypes = " + generateContextTypes(context) + ";\n\n          return Comp;\n        });\n      ", { presets: ["es2015", "react", "stage-1"] }).code;
      } else {
        return (0, _babelStandalone.transform)("\n        ((" + Object.keys(scopeWithProps).join(",") + ", mountNode) => {\n          " + code + "\n        });\n      ", { presets: ["es2015", "react", "stage-1"] }).code;
      }
    }, _this._executeCode = function () {
      var mountNode = _this.mount;
      var _this$props2 = _this.props,
          scope = _this$props2.scope,
          noRender = _this$props2.noRender,
          previewComponent = _this$props2.previewComponent;


      var scopeWithProps = _extends({}, scope, { PropTypes: _propTypes2.default });

      var tempScope = [];

      try {
        Object.keys(scopeWithProps).forEach(function (s) {
          return tempScope.push(scopeWithProps[s]);
        });
        tempScope.push(mountNode);
        var compiledCode = _this._compileCode();
        if (noRender) {
          /* eslint-disable no-eval, max-len */
          var Comp = _react2.default.createElement(eval(compiledCode).apply(undefined, tempScope));
          _server2.default.renderToString(_react2.default.createElement(previewComponent, {}, Comp));
          (0, _reactDom.render)(_react2.default.createElement(previewComponent, {}, Comp), mountNode);
        } else {
          eval(compiledCode).apply(undefined, tempScope);
        }
        /* eslint-enable no-eval, max-len */
        clearTimeout(_this.timeoutID);
        _this.setState({ error: null });
      } catch (err) {
        var error = err.toString();
        clearTimeout(_this.timeoutID); //eslint-disable-line no-undef
        _this.timeoutID = setTimeout(function () {
          _this.setState({ error: error });
        }, 500);
      }
    }, _this.componentDidMount = function () {
      _this._executeCode();
    }, _this.componentDidUpdate = function (prevProps) {
      if (_this.props.code !== prevProps.code) {
        _this._executeCode();
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Preview, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var error = this.state.error;

      return _react2.default.createElement(
        "div",
        null,
        error !== null ? _react2.default.createElement(
          "div",
          { className: "playgroundError" },
          error
        ) : null,
        _react2.default.createElement("div", { ref: function ref(c) {
            _this2.mount = c;
          }, className: "previewArea" })
      );
    }
  }]);

  return Preview;
}(_react.Component);

Preview.defaultProps = {
  previewComponent: "div"
};
Preview.propTypes = {
  code: _propTypes2.default.string.isRequired,
  scope: _propTypes2.default.object.isRequired,
  previewComponent: _propTypes2.default.node,
  noRender: _propTypes2.default.bool,
  context: _propTypes2.default.object
};
exports.default = Preview;
